File: src/com/train/algorithm/list/KWayMerge.java
package com.train.algorithm.list;

import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

// 多路归并
public class KWayMerge {
    class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    // leetcode-23
    class Status implements Comparable<Status>{
        int val;
        ListNode node;

        Status(int val,ListNode node){
            this.val = val;
            this.node = node;
        }

        @Override
        public int compareTo(Status node1){
            return this.val - node1.val;
        }
    }
    PriorityQueue<Status> queue = new PriorityQueue<>();
    public ListNode mergeKLists(ListNode[] lists) {
        for (ListNode n : lists) {
            if (n != null) {
                queue.offer(new Status(n.val,n));
            }
        }
        ListNode head = new ListNode(0);
        ListNode tail = head;
        while(!queue.isEmpty()){
            Status temp = queue.poll();
            tail.next = temp.node;
            tail = tail.next;
            if (temp.node.next != null) {
                queue.offer(new Status(temp.node.next.val,temp.node.next));
            }
        }
        return head.next;
    }

    // leetcode-378
    public int kthSmallest(int[][] matrix, int k) {
        PriorityQueue<int[]> queue = new PriorityQueue<int[]>(new Comparator<int[]>(){
            @Override
            public int compare(int[] a, int[] b){
                return a[0]-b[0];
            }
        });
        int n = matrix.length;
        for (int i = 0;i < n ;i++ ) {
            queue.offer(new int[]{matrix[i][0],i,0});
        }
        for (int i = 0;i < k - 1 ;i++ ) {
            int[] temp = queue.poll();
            if (temp[2] != n - 1) {
                queue.offer(new int[]{matrix[temp[1]][temp[2] + 1],temp[1],temp[2] + 1});
            }
        }
        return queue.poll()[0];
    }

    // leetcode-632
    public int[] smallestRange(List<List<Integer>> nums) {
        int left = 0,right = Integer.MAX_VALUE;
        int min = right - left;
        int max = Integer.MIN_VALUE;
        int size = nums.size();
        int[] next = new int[size];
        PriorityQueue<Integer> queue = new PriorityQueue<Integer>(new Comparator<Integer>(){
            @Override
            public int compare(Integer index1,Integer index2){
                return nums.get(index1).get(next[index1]) - nums.get(index2).get(next[index2]);
            }
        });
        for (int i = 0;i < size ;i++ ) {
            queue.offer(i);
            max = Math.max(max,nums.get(i).get(0));
        }
        while(true){
            int minIndex = queue.poll();
            int cur = max - nums.get(minIndex).get(next[minIndex]);
            if (cur < min) {
                min = cur;
                left = nums.get(minIndex).get(next[minIndex]);
                right = max;
            }
            next[minIndex]++;
            if (next[minIndex] == nums.get(minIndex).size()) {
                break;
            }
            queue.offer(minIndex);
            max = Math.max(max,nums.get(minIndex).get(next[minIndex]));
        }
        return new int[]{left,right};
    }
}

File: src/com/train/algorithm/list/implementInJS/KWayMerge.js
// 多路归并
// leetcode-23
var mergeKLists = function(lists) {
	if(!lists.length){
		return null;
	}
	if(lists.length === 1){
		return lists[0];
	}
	lists.splice(0,2,mergeList(lists[0],lists[1]));
	return mergeKLists(lists);
};

function mergeList(l1,l2){
	let head = new ListNode(),pre = head;
	while(l1 && l2){
		if(l1.val > l2.val){
			pre.next = l2;
			l2 = l2.next;
		}else{
			pre.next = l1;
			l1 = l1.next;
		}
		pre = pre.next;
	}
	pre.next = l1 ? l1 : l2;
	return head.next;
}

// leetcode-378
const countMatrix = (matrix,midVal) => {
	const n = matrix.length;
	let count = 0;
	let row = 0;
	let col = n-1;
	while(row < n && col >= 0){
		if(midVal >= matrix[row][col]){
			count = count + col + 1;
			row++;
		}else{
			col--;
		}
	}
	return count;
}
var kthSmallest = function(matrix, k) {
	const n = matrix.length;
	let low = matrix[0][0];
	let high = matrix[n-1][n-1];
	while(low <= high){
		let midVal = low + ((high - low) >>> 1);
		let count = countMatrix(matrix,midVal);
		if(count < k){
			low = midVal + 1;
		}else{
			high = midVal - 1;
		}
	}
	return low;
};

// leetcode-632
var smallestRange = function(nums) {
	let allNums = [];
	let map = {};
	for (let i = 0;i < nums.length;i++){
		map[i] = 0;
		for(let j = 0;j < nums[i].length;j++){
			allNums.push({
				num : nums[i][j],
				type : i
			});
		}
	}
	allNums.sort((a,b) => {
		return a.num - b.num;
	})
	let left = 0;
	let count = 0;
	let minLen = Infinity;
	let minStart = 0;
	for(let right = 0;right < allNums.length;right++){
		if(map[allNums[right].type] === 0){
			count++;
		}
		map[allNums[right].type]++;
		while(count == nums.length && left <= right){
			if(allNums[right].num - allNums[left].num < minLen){
				minLen = allNums[right].num - allNums[left].num;
				minStart = allNums[left].num;
			}
			map[allNums[left].type]--;
			if(map[allNums[left].type] === 0){
				count--;
			}
			left++;
		}
	}
	return [minStart,minStart + minLen];
};